### PR TITLE
feat(api): Update URL interceptor to be async

### PR DESCRIPTION
**Pull request by royjit**
## Description

- Update `URLRequestInterceptor` with new async protocols
- Implementation of interceptor for token and IAM request
- Added chaining interceptors

## General Checklist

- [ ] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [x] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [x] PR title conforms to conventional commit style
- [ ] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by thisisabhash** [comment]
nit: Could we please have the description and checklist updated in the PR?

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSRESTOperation.swift [hunk 1]
```diff
@@ -92,37 +92,63 @@ final public class AWSRESTOperation: AmplifyOperation<
...
+
+        var mutableIterator = iterator
+        guard let interceptor = mutableIterator.next() else {
+            completion(request)
             return
         }
+        interceptor.intercept(request) { result in
+            do {
+                let interceptedRequest = try result.get()
+                completion(interceptedRequest)
```
Comment (thisisabhash) [inline AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSRESTOperation.swift:135]: should we call `chainInterceptors(mutableIterator, interceptedRequest)` here?

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSGraphQLOperation.swift [hunk 2]
```diff
@@ -100,35 +100,62 @@ final public class AWSGraphQLOperation<R: Decodable>: GraphQLOperation<R> {
...
-        task.resume()
+        var mutableIterator = iterator
+        guard let interceptor = mutableIterator.next() else {
+            completion(request)
+            return
+        }
+        interceptor.intercept(request) { result in
+            do {
+                let interceptedRequest = try result.get()
+                completion(interceptedRequest)
```
Comment (thisisabhash) [inline AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSGraphQLOperation.swift:142]: should we call chainInterceptors(mutableIterator, interceptedRequest) here?

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/RequestInterceptor/AuthTokenURLRequestInterceptor.swift [hunk 3]
```diff
@@ -41,4 +41,29 @@ struct AuthTokenURLRequestInterceptor: URLRequestInterceptor {
         mutableRequest.setValue(token, forHTTPHeaderField: "authorization")
         return mutableRequest as URLRequest
     }
+
+    func intercept(_ request: URLRequest, completion: @escaping (Result<URLRequest, Error>) -> Void) {
+        guard let mutableRequest = (request as NSURLRequest).mutableCopy() as? NSMutableURLRequest else {
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/RequestInterceptor/AuthTokenURLRequestInterceptor.swift:46]: no, the previous implementation uses semaphore logic, which we are trying not to use.

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/RequestInterceptor/IAMURLRequestInterceptor.swift [hunk 4]
```diff
@@ -62,4 +62,49 @@ struct IAMURLRequestInterceptor: URLRequestInterceptor {
...
+
+        let signer: AWSSignatureV4Signer = AWSSignatureV4Signer(
+            credentialsProvider: iamCredentialsProvider.getCredentialsProvider(),
+            endpoint: endpoint)
+
+        signer.interceptRequest(mutableRequest).continueWith { task in
+            if let error = task.error {
+                let apiError = APIError.operationError("Got error trying to sign", "", error)
+                completion(.failure(apiError))
+                return nil
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/RequestInterceptor/IAMURLRequestInterceptor.swift:104]: The signer interceptor is from AWS SDK for iOS which uses AWSTaskCompletion. It require some value to be returned.

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSGraphQLOperation.swift (hunk 2 above)
Comment (royjit) [reply AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSGraphQLOperation.swift:142]: oops! thank you for catching this.

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/RequestInterceptor/AuthTokenURLRequestInterceptor.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (thisisabhash) [reply AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/RequestInterceptor/AuthTokenURLRequestInterceptor.swift:46]: oh got it. i didn't notice the `authProvider.getTokens()` had two flavors.

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSRESTOperation.swift [hunk 5]
```diff
@@ -92,37 +92,66 @@ final public class AWSRESTOperation: AmplifyOperation<
...
             }
+
+            // Begin network task
+            let task = self.session.dataTaskBehavior(with: finalRequest)
+            self.mapper.addPair(operation: self, task: task)
+            task.resume()
         }
+    }
+
+    private func chainInterceptors<I: IteratorProtocol>(
```
Comment (thisisabhash) [inline AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSRESTOperation.swift:116]: nit: it'd be a good idea if we rename this function to something like `handleRequestInChainedInterceptors`

**Review by thisisabhash — APPROVED** [review]
Minor comment for renaming a function. LGTM

**Review by 5d — APPROVED** [review]

**Pull request merged**